pallets/msa/src/lib.rs

```diff
@@ -399,6 +408,51 @@ pub mod pallet {
 		fn on_initialize(current: T::BlockNumber) -> Weight {
 			Self::reset_virtual_bucket_if_needed(current)
 		}
+
+		fn offchain_worker(_block_number: T::BlockNumber) {
+			let filtered_events: Vec<Event<T>> =
+				<frame_system::Pallet<T>>::read_events_no_consensus()
+					.into_iter()
+					.filter_map(|event_record| {
+						let local_event = <T as Config>::RuntimeEvent::from(event_record.event);
+						local_event.try_into().ok()
+					})
+					.filter(|msa_event| {
+						matches!(
+							msa_event,
+							Event::PublicKeyAdded { .. } |
+								Event::PublicKeyDeleted { .. } | Event::MsaCreated { .. }
+						)
+					})
+					.collect();
+			let mut events_by_msa_id: BTreeMap<
+				MessageSourceId,
+				Vec<(T::AccountId, offchain_storage::EventType)>,
+			> = BTreeMap::new();
+
+			for event in filtered_events {
+				match event {
+					Event::MsaCreated { msa_id, key } => {
+						let events = events_by_msa_id.entry(msa_id).or_default();
+						events.push((key, offchain_storage::EventType::Add));
+					},
+					Event::PublicKeyAdded { msa_id, key } => {
+						let events = events_by_msa_id.entry(msa_id).or_default();
+						events.push((key, offchain_storage::EventType::Add));
+					},
+					Event::PublicKeyDeleted { msa_id, key } => {
+						let events = events_by_msa_id.entry(msa_id).or_default();
+						events.push((key, offchain_storage::EventType::Remove));
+					},
+					_ => {},
+				}
+			}
+			if !events_by_msa_id.is_empty() {
+				for (msa_id, events) in events_by_msa_id {
+					Self::process_events(msa_id, events);
```

Comment:
Collect a "replay" of events per msa_id and process on msa_id basis (reason, we lock with msa_id)